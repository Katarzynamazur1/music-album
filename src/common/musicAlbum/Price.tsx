interface PriceProps {
  value: number;
  percentDiscount?: number;
}

const priceWrapperStyle = {
  margin: "8px",
};

interface PriceStyleConfig {
  strikeThrough?: boolean;
  highlighted?: boolean;
}

const getPriceStyle = (config: PriceStyleConfig) => {
  return {
    padding: "0 5px",
    textDecoration: config.strikeThrough ? "line-through" : undefined,
    backgroundColor: config.highlighted ? "yellow" : "inherit",
    color: config.highlighted ? "black" : "inherit",
  };
};

const getPriceAfterDiscount = (
  basePrice: number,
  percentDiscount: number | undefined
) => {
  const priceAfterDiscount = percentDiscount
    ? basePrice - (basePrice * percentDiscount) / 100
    : null;

  return priceAfterDiscount?.toFixed(2);
};

const Price: React.FC<PriceProps> = ({ value, percentDiscount }) => {
  const valueAfterDiscount = getPriceAfterDiscount(value, percentDiscount);

  const basePriceStyle = getPriceStyle({ strikeThrough: !!valueAfterDiscount });
  const discountedPriceStyle = getPriceStyle({
    highlighted: !!valueAfterDiscount,
  });

  return (
    <span style={priceWrapperStyle}>
      <span style={basePriceStyle}>{value}</span>
      <span style={discountedPriceStyle}>{valueAfterDiscount}</span>
    </span>
  );
};

export default Price;
