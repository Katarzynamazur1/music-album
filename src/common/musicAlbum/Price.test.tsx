import { render, RenderResult } from "@testing-library/react";
import Price from "./Price";

describe("Price.tsx", () => {
  it("with discount", () => {
    const renderResult: RenderResult = render(
      <Price value={200} percentDiscount={20} />
    );
    expect(renderResult.container).toMatchSnapshot();
  });

  it("without discount", () => {
    const renderResult: RenderResult = render(<Price value={200} />);
    expect(renderResult.container).toMatchSnapshot();
  });

  it("displayed discount is mathematically correct", () => {
    const renderResult = render(<Price value={100} percentDiscount={19} />);
    const priceAfterDiscount = renderResult.getByTestId("discounted-price");
    expect(priceAfterDiscount).toHaveTextContent("81");
  });
});
