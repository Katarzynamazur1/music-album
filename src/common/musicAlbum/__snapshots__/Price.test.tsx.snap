// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price.tsx with discount 1`] = `
<div>
  <span
    style="margin: 8px;"
  >
    <span
      style="padding: 0px 5px; text-decoration: line-through; background-color: inherit;"
    >
      200
    </span>
    <span
      data-testid="discounted-price"
      style="padding: 0px 5px; background-color: yellow; color: black;"
    >
      160.00
    </span>
  </span>
</div>
`;

exports[`Price.tsx without discount 1`] = `
<div>
  <span
    style="margin: 8px;"
  >
    <span
      style="padding: 0px 5px; background-color: inherit;"
    >
      200
    </span>
    <span
      data-testid="discounted-price"
      style="padding: 0px 5px; background-color: inherit;"
    />
  </span>
</div>
`;
